clar2wasm/src/wasm_generator.rs

```diff
@@ -1847,6 +1844,1697 @@ impl WasmGenerator {
+                                inner.br(block_id);
+                            },
+                            |_| {},
+                        );
+
+                    // Push the `some` indicator onto the stack
+                    then.i32_const(1);
+
+                    // Reserve space in the call stack to hold the deserialized
+                    // string.
+                    let result_offset = self.module.locals.add(ValType::I32);
+                    then.global_get(self.stack_pointer).local_tee(result_offset);
+                    then.i32_const((type_length * 4) as i32)
+                        .binop(BinaryOp::I32Add)
+                        .global_set(self.stack_pointer);
+
+                    // Push the offset of the result onto the stack
+                    then.local_get(result_offset);
+
+                    // Call utf8 to scalar conversion function. It will leave
+                    // return the length of the string.
+                    then.local_get(offset_local)
+                        .i32_const(5)
+                        .binop(BinaryOp::I32Add)
+                        .local_get(string_length)
+                        .local_get(result_offset)
+                        .call(
+                            self.module
+                                .funcs
+                                .by_name("stdlib.convert-utf8-to-scalars")
+                                .expect("stdlib.convert-utf8-to-scalars not found"),
+                        );
+
+                    // Increment the offset by the length of the serialized
+                    // buffer.
+                    then.local_get(computed_end).local_set(offset_local);
+                },
+                |else_| {
+                    // Invalid prefix, return `none`.
+                    else_.i32_const(0).i32_const(0).i32_const(0);
+                },
+            );
+
+        // Add our main block to the builder.
+        builder.instr(walrus::ir::Block { seq: block_id });
+
+        Ok(())
+    }
+
+    /// Deserialize a buffer in memory using the consensus serialization rules.
+    /// The offset and length of the buffer are on the top of the data stack.
+    /// Leaves `(some value)` on the top of the stack, or `none` if
+    /// deserialization fails. It also updates `offset_local` to point to the
+    /// next byte after the bytes used for deserialization. The top-level
+    /// caller of this function should verify that the entire buffer was used
+    /// in deserialization.
+    /// See SIP-005 for deserialization details.
+    pub(crate) fn deserialize_from_memory(
+        &mut self,
+        builder: &mut InstrSeqBuilder,
+        offset_local: LocalId,
+        end_local: LocalId,
+        ty: &TypeSignature,
+    ) -> Result<(), GeneratorError> {
+        let memory = self.get_memory();
+
+        use clarity::vm::types::signatures::TypeSignature::*;
+        match ty {
```

Comment:
Here is a place where it might make more sense to use the trait features of the language, rather than large match statements. (as in the refactor of wasm_generator into words trait)

Reply:
Yeah, I thought about this, but didn't pull the trigger. If I can get to it today, I'll make this change, otherwise, let's merge this and come back to it later.

Reply:
Just to clarify, you'r picturing something like this?

```rust
let deserializer = get_deserializer(ty);
deserializer.deserialize_from_memory(builder, offset, end, &value_ty)?;
```

Reply:
Let's do this refactoring later. For now, I'll just move them into a separate file as-is.